src/lammpsio/snapshot.py

```diff
@@ -171,36 +230,128 @@ def to_hoomd_gsd(self, type_map=None):
+                    angle_label_map = LabelMap(map=angle_label_map)
+                # set angle types & typeid
+                frame.angles.types = angle_label_map.types
+                _set_type_id(
+                    self.angles.typeid,
+                    frame.angles.typeid,
+                    angle_label_map,
+                )
+        if self.has_dihedrals():
+            if self.dihedrals.has_typeid():
+                frame.dihedrals.N = self.dihedrals.N
+                frame.dihedrals.types = self.dihedrals.id
+                frame.dihedrals.group = self.dihedrals.members - 1
+                frame.dihedrals.typeid = numpy.zeros(self.dihedrals.N, dtype=int)
+                if self.dihedrals.type_label is not None:
+                    dihedral_label_map = self.dihedrals.type_label
+                else:
+                    sorted_typeids = numpy.sort(numpy.unique(self.dihedrals.typeid))
+                    dihedral_label_map = {}
+                    for typeid in sorted_typeids:
+                        dihedral_label_map[typeid] = str(typeid)
+                    dihedral_label_map = LabelMap(map=dihedral_label_map)
+                # set dihedral types & typeid
+                frame.dihedrals.types = dihedral_label_map.types
+                _set_type_id(
+                    self.dihedrals.typeid,
+                    frame.dihedrals.typeid,
+                    dihedral_label_map,
+                )
+        if self.has_impropers():
+            if self.impropers.has_typeid():
+                frame.impropers.N = self.impropers.N
+                frame.impropers.types = self.impropers.id
+                frame.impropers.group = self.impropers.members - 1
+                frame.impropers.typeid = numpy.zeros(self.impropers.N, dtype=int)
+                if self.impropers.type_label is not None:
+                    improper_label_map = self.impropers.type_label
+                else:
+                    sorted_typeids = numpy.sort(numpy.unique(self.impropers.typeid))
+                    improper_label_map = {}
+                    for typeid in sorted_typeids:
+                        improper_label_map[typeid] = str(typeid)
+                    improper_label_map = LabelMap(map=improper_label_map)
+                # set improper types & typeid
+                frame.impropers.types = improper_label_map.types
+                _set_type_id(
+                    self.impropers.typeid,
+                    frame.impropers.typeid,
+                    improper_label_map,
+                )
 
         # undo the sort so object goes back the way it was
         if reverse_order is not None:
             self.reorder(reverse_order, check_order=False)
 
-        if (
-            self.has_bonds()
-            or self.has_angles()
-            or self.has_dihedrals()
-            or self.has_impropers()
-        ):
-            warnings.warn("Conversion of topology to gsd is not supported")
-
+        frame.validate()
```

Comment:
I don't think a validation is necessary when going _to_ GSD. Maybe in the _from_ direction to sanitize our inputs.